pkg/identity/identity.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2019, Oath Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package identity
+
+import (
+	"net"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/yahoo/athenz/libs/go/zmssvctoken"
+	"github.com/yahoo/k8s-athenz-syncer/pkg/crypto"
+)
+
+var (
+	gracePeriod = 5 * time.Minute
+)
+
+// PrivateKeyProvider returns a signing key on demand.
+type PrivateKeyProvider func() (*crypto.SigningKey, error)
+
+// tokenProvider provides tokens.
+type tokenProvider interface {
+	Token() (string, error)
+}
+
+// IdentityProvider provides ntokens and optional TLS certs.
+type IdentityProvider struct {
+	tokenProvider
+	domain  string
+	service string
+	podIP   net.IP
+	ks      PrivateKeyProvider
+}
+
+// tp implements tokenProvider
+type tp struct {
+	domain  string
+	service string
+	ip      string
+	host    string
+	expiry  time.Duration
+	ks      PrivateKeyProvider
+	l       sync.Mutex
+	current string
+	expire  time.Time
+}
+
+// Config is the identity provider configuration.
+type Config struct {
+	Domain                    string             // Athenz domain
+	Service                   string             // Athenz service
+	PodIP                     net.IP             // Pod IP for IP SAN, nil ok for now
+	PrivateKeyProvider        PrivateKeyProvider // source for private keys
+	TokenHost                 string             // ntoken hostname where available
+	TokenIP                   net.IP             // ntoken IP address
+	TokenExpiry               time.Duration      // token expire time
+	X509CertFile              string             // file path for x509 cert
+	X509KeyFile               string             // file path for x509 key
+	AthenzClientAuthnx509Mode bool               // athenz client authentication. default is token.
+}
+
+// NewIdentityProvider returns an identity provider for the supplied configuration.
+// It returns either x509Provider or TokenProvider based on the presence of x509 configurations.
+func NewIdentityProvider(config Config) (*IdentityProvider, error) {
+	if config.TokenExpiry == 0 {
+		config.TokenExpiry = time.Hour
+	}
+	var ipStr string
+	if config.TokenIP != nil {
+		ipStr = config.TokenIP.String()
+	}
+
+	ip := &IdentityProvider{
```

Comment:
if we eliminate the IdentityProvider struct, and probably have only TokenProvider {}, would that simplify a bit?